addons/crm/models/res_partner.py

```diff
@@ -16,9 +16,10 @@ class Partner(models.Model):
 
     @api.multi
     def _compute_opportunity_count(self):
+        partners_data = self.env['crm.lead'].read_group([('partner_id', 'in', self.ids), ('type', '=', 'opportunity')], ['partner_id'], ['partner_id'])
+        mapped_data = dict([(partner['partner_id'][0], partner['partner_id_count']) for partner in partners_data])
         for partner in self:
-            operator = 'child_of' if partner.is_company else '='  # the opportunity count should counts the opportunities of this company and all its contacts
-            partner.opportunity_count = self.env['crm.lead'].search_count([('partner_id', operator, partner.id), ('type', '=', 'opportunity')])
+            partner.opportunity_count = mapped_data.get(partner.id, 0) + sum(mapped_data.get(int(child), 0) for child in partner.mapped('child_ids'))
```

Comment:
I don't think this code is correct. 
So here, we are summing the children count to the one from the parent. The parent (current partner in the loop) is obviously present in the `mapped_data` dict as the read_group domain takes ` self.ids`. But, the children might not be in `self.ids`, and so will not be present in the dict.
If you have tested and it was working, i guess the children was on your kanban view (for instance). I believe, if you tried with a domain excluding some children, the count of a parent partner will not be correct.